core/src/test/java/kafka/testkit/TestKitNodes.java

```diff
@@ -119,7 +101,35 @@ public TestKitNodes build() {
             if (bootstrapMetadataVersion == null) {
                 bootstrapMetadataVersion = MetadataVersion.latest();
             }
-            return new TestKitNodes(clusterId, bootstrapMetadataVersion, controllerNodes, brokerNodes);
+            String baseDirectory = TestUtils.tempDirectory("kafka_" + clusterId).getAbsolutePath();
+            try {
+                NavigableMap<Integer, ControllerNode> controllerNodes = new TreeMap<>();
+                for (ControllerNode.Builder controllerNodeBuilder : controllerNodeBuilders.values()) {
+                    ControllerNode controllerNode = controllerNodeBuilder.build(baseDirectory);
+                    if (controllerNodes.put(controllerNode.id(), controllerNode) != null) {
+                        throw new RuntimeException("More than one controller claimed ID " + controllerNode.id());
+                    }
+                }
+                NavigableMap<Integer, BrokerNode> brokerNodes = new TreeMap<>();
+                for (BrokerNode.Builder brokerNodeBuilder : brokerNodeBuilders.values()) {
+                    BrokerNode brokerNode = brokerNodeBuilder.build(baseDirectory);
+                    if (brokerNodes.put(brokerNode.id(), brokerNode) != null) {
+                        throw new RuntimeException("More than one broker claimed ID " + brokerNode.id());
+                    }
+                }
+                return new TestKitNodes(baseDirectory,
+                    clusterId,
+                    bootstrapMetadataVersion,
+                    controllerNodes,
+                    brokerNodes);
+            } catch (Exception e) {
```

Comment:
nit: can be `catch (RuntimeException e) {` since the code doesn't throw any checked exceptions and the signature of this method does not declare any checked exceptions as being potentially thrown.

Reply:
Yeah, that's fair. We can only have `RuntimException` here. Still, I can see no logical reason why we wouldn't do the deletion if the code was restructured to throw checked exceptions here. So it seems clearer this way.